### PR TITLE
Fix return code for examples under programs/

### PR DESCRIPTION
**Pull request by andresag01**
Posix states that exit codes can only be 8-bits. This commit modifies
the examples under programs/ to ensure that whenever there is a failure
the integer 1 is returned, otherwise 0. Programs that already had this
check in place were not modified.

### CODE REVIEW

**Comment by andresag01** [comment]
This pull request addresses:
  -  https://github.com/ARMmbed/mbedtls/issues/412
  -  https://github.com/ARMmbed/mbedtls/issues/352

**Comment by andresag01** [comment]
@sbutcher-arm @yanesca

File: programs/aes/aescrypt2.c [hunk 1]
```diff
@@ -447,6 +447,9 @@ int main( int argc, char *argv[] )
     mbedtls_aes_free( &aes_ctx );
     mbedtls_md_free( &sha_ctx );
 
+    if( ret != 0 && ret != 1)
```
Comment (yanesca) [inline programs/aes/aescrypt2.c]: It probably doesn't matter, but wouldn't be `if( ret != 0 )` sufficient?

File: programs/aes/aescrypt2.c [hunk 2]
```diff
@@ -447,8 +452,8 @@ int main( int argc, char *argv[] )
     mbedtls_aes_free( &aes_ctx );
     mbedtls_md_free( &sha_ctx );
 
-    if( ret != 0 && ret != 1)
-        ret = 1;
+    if( ret != MBEDTLS_EXIT_SUCCESS && ret != MBEDTLS_EXIT_FAILURE )
```
Comment (yanesca) [inline programs/aes/aescrypt2.c]: It probably doesn't matter, but wouldn't be `if( ret != MBEDTLS_EXIT_SUCCESS )` sufficient?

**Comment by sbutcher-arm** [comment]
In many of the examples we're using `ret` to take the result code from a library function and then using `ret` again to pass back an exit code. `ret` can take any value of the function, and yet the exit code can only be `MBEDTLS_EXIT_SUCCESS` or `MBEDTLS_EXIT_FAILURE`.

Can you please separate these two uses into two different variables? I suggest `result` and `exit code` and remove `ret` entirely. The change will help find any missed `ret` variables as it won't build.

**Comment by sbutcher-arm** [comment]
Ok, thank you for reworking this, but it's not quite what I meant.

In programs/aes/aescrypt2.c, line 106 and line 112 we're using `ret` for two different purposes. Once, to store the return value of `mbedtls_md_setup()` and again as the exit code. Those two uses should be done with separate variables.

This mistake is repeated again and again through many example apps. Could you please use something like the following design pattern:
```
int retval = 0;
int exitcode = MBEDTLS_EXIT_SUCCESS;

// Do some stuff

retval = mbedtls_some_fn();
if( retval != 0 )
{
    exitcode = MBEDTLS_EXIT_FAILURE;
    goto cleanup;
}
```

 Could you please fix it?

File: programs/aes/aescrypt2.c [hunk 3]
```diff
@@ -105,6 +108,8 @@ int main( int argc, char *argv[] )
     {
         mbedtls_printf( "  ! mbedtls_md_setup() returned -0x%04x\n", -ret );
         goto exit;
+    } else {
```
Comment (sbutcher-arm) [inline programs/aes/aescrypt2.c]: This else and resetting `ret` would not be necessary if you used a separate variable for the exit code and for the return value from `mbedtls_md_setup()`.

**Comment by pjbakker** [comment]
Simon,

I concur that separate values make sense here.

I don't fully agree with your proposal.

Defensive programming suggests the initial value should be failure and it should only be set to success on correct traversal of the app, instead of the other way around.


-------- Original message --------

**Comment by sbutcher-arm** [comment]
@pjbakker, yes, you're right. I was trying to show how to restructure the code and I didn't think too hard about that as I wrote it.

**Comment by andresag01** [comment]
@sbutcher-arm: modified PR according to comments.

File: programs/aes/crypt_and_hash.c [hunk 4]
```diff
@@ -172,7 +176,7 @@ int main( int argc, char *argv[] )
         mbedtls_fprintf( stderr, "Cipher '%s' not found\n", argv[4] );
         goto exit;
     }
-    if( ( ret = mbedtls_cipher_setup( &cipher_ctx, cipher_info) ) != 0 )
+    if( ( retval = mbedtls_cipher_setup( &cipher_ctx, cipher_info) ) != 0 )
```
Comment (sbutcher-arm) [inline programs/aes/crypt_and_hash.c]: You're making an assignment to retval here, yet the variable is never referenced again.

I suggest rather than just remove the variable, split the assignment to `retval` and the test of the return value across two lines (easier to read), then output the value of `retval` if it fails.

File: programs/pkey/dh_client.c [hunk 5]
```diff
@@ -113,18 +117,18 @@ int main( void )
 
     if( ( f = fopen( "rsa_pub.txt", "rb" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/dh_client.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/dh_client.c [hunk 6]
```diff
@@ -173,23 +177,23 @@ int main( void )
...
     {
-        mbedtls_printf( " failed\n  ! mbedtls_dhm_read_params returned %d\n\n", ret );
+        mbedtls_printf( " failed\n  ! mbedtls_dhm_read_params returned %d\n\n", retval );
         goto exit;
     }
 
     if( dhm.len < 64 || dhm.len > 512 )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/dh_client.c]: Again, why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/dh_client.c [hunk 7]
```diff
@@ -205,17 +209,17 @@ int main( void )
 
     if( ( n = (size_t) ( end - p ) ) != rsa.len )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/dh_client.c]: Again, why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/dh_genprime.c [hunk 8]
```diff
@@ -163,22 +167,24 @@ int main( int argc, char **argv )
 
     if( ( fout = fopen( "dh_prime.txt", "wb+" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/dh_genprime.c]: Again, why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/dh_server.c [hunk 9]
```diff
@@ -114,24 +118,24 @@ int main( void )
 
     if( ( f = fopen( "rsa_priv.txt", "rb" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/dh_server.c]: Again, why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/dh_server.c [hunk 10]
```diff
@@ -148,7 +152,7 @@ int main( void )
 
     if( ( f = fopen( "dh_prime.txt", "rb" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/dh_server.c]: Again, why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/gen_key.c [hunk 11]
```diff
@@ -206,7 +210,7 @@ int main( int argc, char *argv[] )
     if( argc == 0 )
     {
     usage:
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/gen_key.c]: Again, why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/key_app_writer.c [hunk 12]
```diff
@@ -204,7 +208,7 @@ int main( int argc, char *argv[] )
     if( argc == 0 )
     {
     usage:
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/key_app_writer.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_genkey.c [hunk 13]
```diff
@@ -101,14 +105,14 @@ int main( void )
     if( ( fpub = fopen( "rsa_pub.txt", "wb+" ) ) == NULL )
     {
         mbedtls_printf( " failed\n  ! could not open rsa_pub.txt for writing\n\n" );
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_genkey.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_genkey.c [hunk 14]
```diff
@@ -118,20 +122,20 @@ int main( void )
     if( ( fpriv = fopen( "rsa_priv.txt", "wb+" ) ) == NULL )
     {
         mbedtls_printf( " failed\n  ! could not open rsa_priv.txt for writing\n" );
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_genkey.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_sign.c [hunk 15]
```diff
@@ -55,15 +58,16 @@ int main( void )
...
+    int exitcode = MBEDTLS_EXIT_FAILURE;
     size_t i;
     mbedtls_rsa_context rsa;
     unsigned char hash[32];
     unsigned char buf[MBEDTLS_MPI_MAX_SIZE];
     char filename[512];
 
     mbedtls_rsa_init( &rsa, MBEDTLS_RSA_PKCS_V15, 0 );
-    ret = 1;
+    retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_sign.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_sign.c [hunk 16]
```diff
@@ -81,22 +85,22 @@ int main( int argc, char *argv[] )
 
     if( ( f = fopen( "rsa_priv.txt", "rb" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_sign.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_sign.c [hunk 17]
```diff
@@ -142,7 +146,7 @@ int main( int argc, char *argv[] )
 
     if( ( f = fopen( filename, "wb+" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_sign.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_sign_pss.c [hunk 18]
```diff
@@ -88,28 +92,28 @@ int main( int argc, char *argv[] )
...
     }
 
     mbedtls_printf( "\n  . Reading private key from '%s'", argv[1] );
     fflush( stdout );
 
-    if( ( ret = mbedtls_pk_parse_keyfile( &pk, argv[1], "" ) ) != 0 )
+    if( ( retval = mbedtls_pk_parse_keyfile( &pk, argv[1], "" ) ) != 0 )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_sign_pss.c]: Initialising `retval`here is corrupting the error code that will be displayed at line 110! Please remove!

File: programs/pkey/rsa_sign_pss.c [hunk 19]
```diff
@@ -88,28 +92,28 @@ int main( int argc, char *argv[] )
...
         mbedtls_printf( " failed\n  ! Could not read key from '%s'\n", argv[1] );
-        mbedtls_printf( "  ! mbedtls_pk_parse_public_keyfile returned %d\n\n", ret );
+        mbedtls_printf( "  ! mbedtls_pk_parse_public_keyfile returned %d\n\n", retval );
         goto exit;
     }
 
     if( !mbedtls_pk_can_do( &pk, MBEDTLS_PK_RSA ) )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_sign_pss.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_sign_pss.c [hunk 20]
```diff
@@ -145,7 +149,7 @@ int main( int argc, char *argv[] )
 
     if( ( f = fopen( filename, "wb+" ) ) == NULL )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_sign_pss.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_verify.c [hunk 21]
```diff
@@ -100,7 +103,7 @@ int main( int argc, char *argv[] )
     /*
      * Extract the RSA signature from the text file
      */
-    ret = 1;
+    retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_verify.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_verify_pss.c [hunk 22]
```diff
@@ -82,16 +86,16 @@ int main( int argc, char *argv[] )
...
         mbedtls_printf( " failed\n  ! Could not read key from '%s'\n", argv[1] );
-        mbedtls_printf( "  ! mbedtls_pk_parse_public_keyfile returned %d\n\n", ret );
+        mbedtls_printf( "  ! mbedtls_pk_parse_public_keyfile returned %d\n\n", retval );
         goto exit;
     }
 
     if( !mbedtls_pk_can_do( &pk, MBEDTLS_PK_RSA ) )
     {
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_verify_pss.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/pkey/rsa_verify_pss.c [hunk 23]
```diff
@@ -101,7 +105,7 @@ int main( int argc, char *argv[] )
     /*
      * Extract the RSA signature from the file
      */
-    ret = 1;
+    retval = 1;
```
Comment (sbutcher-arm) [inline programs/pkey/rsa_verify_pss.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/ssl/ssl_mail_client.c [hunk 24]
```diff
@@ -521,41 +524,41 @@ int main( int argc, char *argv[] )
...
     else
 #endif
 #if defined(MBEDTLS_CERTS_C)
-        ret = mbedtls_x509_crt_parse( &clicert, (const unsigned char *) mbedtls_test_cli_crt,
+        retval = mbedtls_x509_crt_parse( &clicert, (const unsigned char *) mbedtls_test_cli_crt,
                               mbedtls_test_cli_crt_len );
 #else
     {
-        ret = -1;
+        retval = -1;
```
Comment (sbutcher-arm) [inline programs/ssl/ssl_mail_client.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/ssl/ssl_mail_client.c [hunk 25]
```diff
@@ -521,41 +524,41 @@ int main( int argc, char *argv[] )
...
     else
 #endif
 #if defined(MBEDTLS_CERTS_C) && defined(MBEDTLS_PEM_PARSE_C)
-        ret = mbedtls_pk_parse_key( &pkey, (const unsigned char *) mbedtls_test_cli_key,
+        retval = mbedtls_pk_parse_key( &pkey, (const unsigned char *) mbedtls_test_cli_key,
                 mbedtls_test_cli_key_len, NULL, 0 );
 #else
     {
-        ret = -1;
+        retval = -1;
```
Comment (sbutcher-arm) [inline programs/ssl/ssl_mail_client.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/x509/cert_req.c [hunk 26]
```diff
@@ -155,7 +159,7 @@ int main( int argc, char *argv[] )
     {
     usage:
         mbedtls_printf( USAGE );
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/x509/cert_req.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/x509/cert_write.c [hunk 27]
```diff
@@ -221,7 +225,7 @@ int main( int argc, char *argv[] )
     {
     usage:
         mbedtls_printf( USAGE );
-        ret = 1;
+        retval = 1;
```
Comment (sbutcher-arm) [inline programs/x509/cert_write.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

File: programs/x509/cert_write.c [hunk 28]
```diff
@@ -503,7 +507,7 @@ int main( int argc, char *argv[] )
                          &mbedtls_pk_rsa( *issuer_key )->E ) != 0 )
         {
             mbedtls_printf( " failed\n  !  issuer_key does not match issuer certificate\n\n" );
-            ret = -1;
+            retval = -1;
```
Comment (sbutcher-arm) [inline programs/x509/cert_write.c]: Why is `retval` set to 1 here? Looks redundant and should be removed.

**Comment by sbutcher-arm** [comment]
Ok, I've reviewed the entire changeset and there are some common minor mistakes:

 * `retval` is commonly initialised at the top of functions unnecessarily because it's first use is to record the return value of a function. This is a minor inefficiency.
 * `retval` is also being used in expressions to record the value of a function, but in the same expression is also being tested, and the value stored in `retval` is not referenced again. Again, a minor inefficiency.
 * And, again, and again, `retval` is assigned a value unnecessarily as a hangover from it's previous use when we were returning `ret` as the exit code! A minor inefficiency, but also distracting.

File: programs/ssl/ssl_mail_client.c (hunk 24 above)
Comment (andresag01) [inline programs/ssl/ssl_mail_client.c]: This is not redundant, the retval is checked in line 539 to ensure that an error did not occur at any part of the `if` statement.

File: programs/ssl/ssl_mail_client.c (hunk 25 above)
Comment (andresag01) [inline programs/ssl/ssl_mail_client.c]: Same as previous comment

**Comment by andresag01** [comment]
@sbutcher-arm: Reworked PR

**Comment by sbutcher-arm** [comment]
@Patater or @hanno-arm, can you review this so we can merge this? It's a bit overdue.

**Comment by hanno-arm** [comment]
@sbutcher-arm Yes, I'll look into it.

**Comment by andresag01** [comment]
Rebased on top of development to fix conflicts

**Comment by andresag01** [comment]
@hanno-arm:
>In general, though, I probably miss something but I don't follow why there has been the renaming ret -> retval everywhere.

The rename from `ret` to `retval` is because I was following @sbutcher-arm's suggestion.

>Some programs aren't yet adapted to the new return scheme

This PR is trying to address the problem reported by one of our users that we are not returning proper 8-bit posix codes to the environment. If the program already returns 0 or 1 this is a good posix code, so there is no need to change it.

>The program pkey/key_app_writer has been adapted, but it is missing the assignment exitcode = MBEDTLS_EXIT_SUCCESS on succes; currently it always returns failure.

I will fix that.

File: programs/aes/aescrypt2.c (hunk 1 above)
Comment (Patater) [reply programs/aes/aescrypt2.c]: Yeah, it doesn't matter, and `if( ret != 0)` would be sufficient. It looks a bit funny to have both `!=0` and `!=1`, but it'll work fine. I've no issue to raise here.

However, instead of `1`, we should use `EXIT_FAILURE` in all cases where `1` is destined to be used directly by `exit()`.

File: programs/aes/aescrypt2.c [hunk 29]
```diff
@@ -29,8 +29,11 @@
 #include "mbedtls/platform.h"
 #else
 #include <stdio.h>
-#define mbedtls_fprintf    fprintf
-#define mbedtls_printf     printf
+#include <stdlib.h>
+#define mbedtls_fprintf      fprintf
+#define mbedtls_printf       printf
+#define MBEDTLS_EXIT_SUCCESS EXIT_SUCCESS
```
Comment (Patater) [inline programs/aes/aescrypt2.c]: Why do we wrap `EXIT_SUCCESS` and `EXIT_FAILURE`?

File: programs/aes/aescrypt2.c [hunk 30]
```diff
@@ -457,9 +456,6 @@ int main( int argc, char *argv[] )
     mbedtls_aes_free( &aes_ctx );
     mbedtls_md_free( &sha_ctx );
 
-    if( ret != MBEDTLS_EXIT_SUCCESS && ret != MBEDTLS_EXIT_FAILURE )
-        ret = MBEDTLS_EXIT_FAILURE;
-
-    return( ret );
+    return( exitcode );
```
Comment (Patater) [inline programs/aes/aescrypt2.c]: Should we use `exit()` here?

File: programs/aes/aescrypt2.c [hunk 31]
```diff
@@ -74,7 +74,8 @@ int main( void )
 #else
 int main( int argc, char *argv[] )
 {
-    int ret = MBEDTLS_EXIT_FAILURE;
+    int retval = 0;
```
Comment (Patater) [inline programs/aes/aescrypt2.c]: I would prefer we left this as `ret` instead of `retval`. `ret` is idiomatic for your use case. (It would also have the benefit of making the diff far less noisy.)

File: programs/ssl/ssl_mail_client.c [hunk 32]
```diff
@@ -493,25 +494,26 @@ int main( int argc, char *argv[] )
 
 #if defined(MBEDTLS_FS_IO)
     if( strlen( opt.ca_file ) )
-        ret = mbedtls_x509_crt_parse_file( &cacert, opt.ca_file );
+        retval = mbedtls_x509_crt_parse_file( &cacert, opt.ca_file );
     else
-#endif
+#endif /* MBEDTLS_FS_IO */
```
Comment (Patater) [inline programs/ssl/ssl_mail_client.c]: Don't add this comment here, at least not in this commit. Our style guide only requires the `#endif` comment when the body of the `#if` is large or complicated.

File: programs/ssl/ssl_mail_client.c [hunk 33]
```diff
@@ -493,25 +494,26 @@ int main( int argc, char *argv[] )
...
+                                mbedtls_test_cas_pem_len );
 #else
     {
-        ret = 1;
+        retval = 1;
         mbedtls_printf("MBEDTLS_CERTS_C and/or MBEDTLS_PEM_PARSE_C not defined.");
     }
-#endif
-    if( ret < 0 )
+#endif /* MBEDTLS_CERTS_C && MBEDTLS_PEM_PARSE_C */
```
Comment (Patater) [inline programs/ssl/ssl_mail_client.c]: Please do style changes in a separate commit from semantic changes.

**Comment by hanno-arm** [comment]
@andresag01 @Patater If we start cleaning up the ways error codes are handled and relate to return values, I think we should try to be entirely uniform. For example, as mentioned in the recent #1142, e.g. [`pk_sign`](https://github.com/ARMmbed/mbedtls/blob/development/programs/pkey/pk_sign.c) overwrites error codes by `1` before it calls `mbedtls_strerror` to give a human understable description - this doesn't make sense. My suggestion would be:
1. Never overwrite `ret` to `0 / 1`, to be able to call `mbedtls_strerror` for meaningful error messages.
2. Use a separate exit value variable which carries `EXIT_FAILURE` by default and gets set to `EXIT_SUCCESS` when everything went well. In particular, there is no direct correlation between `ret != 0` being found and a modification of `exitcode`. 
3. Omit the renaming of `ret` to `retval` to keep the PR easier to overlook.

**Comment by andresag01** [comment]
@Patater, @hanno-arm, @sbutcher-arm: I have reworked the changes to make the return value and exit consistent. However, this PR does not address any other issues such as: style, unchecked return codes, etc in the examples. Also, I did not change the following examples: `ssl/mini_client`, `ssl/ssl_client2` and `ssl/dtls_client`.

File: programs/aes/aescrypt2.c [hunk 34]
```diff
@@ -29,9 +29,12 @@
 #include "mbedtls/platform.h"
 #else
 #include <stdio.h>
-#define mbedtls_fprintf    fprintf
-#define mbedtls_printf     printf
-#endif
+#include <stdlib.h>
+#define mbedtls_fprintf         fprintf
+#define mbedtls_printf          printf
+#define MBEDTLS_EXTI_SUCCESS    EXIT_SUCCESS
```
Comment (Patater) [inline programs/aes/aescrypt2.c]: `MBEDTLS_EXTI_SUCCESS`? Why not `MBEDTLS_EXIT_SUCCESS`? This needs fixing in many places. How does this build? What testing has been done?

Comment (andresag01) [reply programs/aes/aescrypt2.c]: >MBEDTLS_EXTI_SUCCESS? Why not MBEDTLS_EXIT_SUCCESS? This needs fixing in many places. 

Sorry about that. You are right, there is a typo here and the mistake is repeated in all the other files because I largely copy-pasted the same definitions across the examples.

> How does this build?

By default, MBEDTLS_PLATFORM_C is defined, so this code is almost never used.

> What testing has been done?

I tested each of the applications with the default config.h that has MBEDTLS_PLATFORM_C so this code was never used. When MBEDTLS_PLATFORM_C is defined, the applications compiled without problems. I mainly tested the application's functionality, but this was in no way thorough as there are 40+ examples. Also, I am not aware of any automated tests to check the examples...

Now that I think about it, I am not even sure why the examples need these defines. I think platform.h defines these macros to something regardless of whether MBEDTLS_PLATFORM_C is or isnt defined and the default values are pretty sensible e.g:

[mbedtls_printf -> printf](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/platform.h#L191)
MBEDTLS_EXIT_SUCCESS -> 0
MBEDTLS_EXIT_FAILURE -> 1

Also, I think these default values for the macros need to be correct, otherwise tests like [this](https://github.com/ARMmbed/mbedtls/blob/development/tests/scripts/all.sh#L593) would probably fail as the library also has dependencies on these standard library symbols. @Patater, I will try replacing this code for `#include "mbedtls/platform.h"`...

**Comment by andresag01** [comment]
@Patater: I have reworked the PR to address the comments.

**Comment by sbutcher-arm** [comment]
@hanno-arm / @Patater - Can you please re-review, based on the rework?

Also, I want this merged soon, but I also want further testing, so am applying the 'do not merge' label for the moment, while we progress the reviews.

**Comment by andresag01** [comment]
@sbutcher-arm, @Patater : Thanks for reviewing the PR. At the moment we have build-only tests in `tests/scripts/all.sh`, but this is clearly not sufficient. Would you like me to add some run tests as part of this PR? As a minimum we should have a sanity test that runs each of the apps and checks the return code is correct and the output look like acceptable...

**Comment by Patater** [comment]
Such a test would be welcome, and since we have to test all these programs anyway as part of getting the PR merged, likely to save time now and in the future if its automated.

I wouldn't block this PR for lack of such an automated test, and would propose if you do make an automated test to help you run the tests required to get this PR merged, that the new automated test go into its own PR.

**Comment by sbutcher-arm** [comment]
@andresag01 - Please hold fire on doing any more work on this PR and doing a test script. The PR is needed, but as you know we also have plenty of other work to do.

**Comment by andresag01** [comment]
@sbutcher-arm: No problem. I will not do any more work on the example testing for now.

**Comment by andresag01** [comment]
@sbutcher-arm: I have just rebased the PR on top of the current development. For reference, when I was implementing the changes, I tested  (most of) the applications one by one and with basic input. However, I did not make any additional testing effort or checked negative tests, etc.

**Comment by andresag01** [comment]
This PR has been reviewed and approved by three separate reviewers, so removing the 'needs review' label

**Comment by sbutcher-arm** [comment]
@andresag01 / @k-stachowiak - I'd like to know that each program has been run at least once, just to test it in one positive test case before we merge.

Andres - can you clarify what testing you've done on this PR, *since* the rebase? It may be possible to skip this step.

**Comment by k-stachowiak** [comment]
@sbutcher-arm I confirm that I have ran each of the program for a single positive test case.

**Comment by andresag01** [comment]
@sbutcher-arm: I have also run almost all the example applications by hand on the simple positive case.

**Comment by andresag01** [comment]
@sbutcher-arm: There are three approvals and tests have been confirmed, could we merge this?

**Comment by sbutcher-arm** [comment]
@andresag01 - Yes! Finally!

**Comment by sbutcher-arm** [comment]
I've backported this fix to the `mbedtls-2.7` branch as PR #1848, but given `mbedtls-2.1` doesn't have defined values for `MBEDTLS_EXIT_FAILURE` and `MBEDTLS_EXIT_SUCCESS`, I don't think it's necessary to backport the PR to `mbedtls-2.1`.
